### PR TITLE
Application start and kill

### PR DESCRIPTION
**Pull request by RomanLevkovych**
Moved sycalls related to starting and killing of applications to separate class, injected to EM.

### CODE REVIEW

File: applications/ExecutionManager/src/application_handler.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+{
+
+using std::string;
+using std::runtime_error;
+
+ApplicationHandler::ApplicationHandler(std::string path)
+        : corePath{std::move(path)}
+{ }
+
+pid_t ApplicationHandler::startProcess(const ProcessInfo &process)
```
Comment (rBekesh) [inline applications/ExecutionManager/src/application_handler.cpp]: (const ProcessInfo& process)

File: applications/ExecutionManager/src/application_handler.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  }
+  return processId;
+}
+
+void ApplicationHandler::killApplication(pid_t processId)
+{
+  kill(processId, SIGTERM);
+}
+
+std::vector<std::string> ApplicationHandler::getArgumentsList(const ProcessInfo &process) const
```
Comment (rBekesh) [inline applications/ExecutionManager/src/application_handler.cpp]: (const ProcessInfo& process)

File: applications/ExecutionManager/src/application_handler.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  std::transform(process.startOptions.cbegin(),
+                 process.startOptions.cend(),
+                 std::back_inserter(arguments),
+                 [](const StartupOption& option)
+                 { return option.makeCommandLineOption(); });
+
+  return arguments;
+}
+
+std::vector<char *> ApplicationHandler::convertToNullTerminatingArgv(std::vector<std::string> &vectorToConvert) const
```
Comment (rBekesh) [inline applications/ExecutionManager/src/application_handler.cpp]: (std::vector<std::string>& vectorToConvert)

File: applications/ExecutionManager/src/application_handler.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+{
+  std::vector<char *> outputVector;
```
Comment (rBekesh) [inline applications/ExecutionManager/src/application_handler.cpp]: <char*>

File: applications/ExecutionManager/src/application_handler.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
+#include "application_handler.hpp"
+
+#include <csignal>
+#include <string>
```
Comment (rBekesh) [inline applications/ExecutionManager/src/application_handler.cpp]: no needed

File: applications/ExecutionManager/src/application_handler.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+#include <exception>
+
+namespace ExecutionManager
+{
+
+using std::string;
```
Comment (rBekesh) [inline applications/ExecutionManager/src/application_handler.cpp]: remove

File: applications/ExecutionManager/src/application_handler.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+using std::runtime_error;
```
Comment (rBekesh) [inline applications/ExecutionManager/src/application_handler.cpp]: remove

File: applications/ExecutionManager/include/iapplication_handler.hpp [hunk 8]
```diff
@@ -0,0 +1,33 @@
+#ifndef IAPPLICATION_HANDLER_HPP
+#define IAPPLICATION_HANDLER_HPP
+
+#include "manifests.hpp"
+
+#include <string>
+#include <vector>
```
Comment (rBekesh) [inline applications/ExecutionManager/include/iapplication_handler.hpp]: not needed

File: applications/ExecutionManager/include/iapplication_handler.hpp [hunk 9]
```diff
@@ -0,0 +1,33 @@
+#ifndef IAPPLICATION_HANDLER_HPP
+#define IAPPLICATION_HANDLER_HPP
+
+#include "manifests.hpp"
+
+#include <string>
```
Comment (rBekesh) [inline applications/ExecutionManager/include/iapplication_handler.hpp]: not needed

File: applications/ExecutionManager/include/iapplication_handler.hpp [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+class IApplicationHandler {
+public:
+    /**
+     * @brief Starts new process with params and returns its' pid
+     * @param application: path to executable
+     * @param arguments: list of arguments (without nullptr)
+     * @return pid of process
+     */
+    virtual pid_t startProcess(const ProcessInfo &application) = 0;
```
Comment (rBekesh) [inline applications/ExecutionManager/include/iapplication_handler.hpp]: (ProcessInfo& application)

File: applications/ExecutionManager/include/iapplication_handler.hpp [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+    /**
+     * @brief Send `signal` to application with provided pid
```
Comment (rBekesh) [inline applications/ExecutionManager/include/iapplication_handler.hpp]: 'sigterm'

File: applications/ExecutionManager/include/application_handler.hpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
+#ifndef EXECUTION_MANAGER_APPLICATION_HANDLER_HPP
+#define EXECUTION_MANAGER_APPLICATION_HANDLER_HPP
+
+#include "iapplication_handler.hpp"
+#include "manifests.hpp"
+
```
Comment (rBekesh) [inline applications/ExecutionManager/include/application_handler.hpp]: #include <string>
#include <vector>

File: applications/ExecutionManager/include/application_handler.hpp [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+namespace ExecutionManager
+{
+
+class ApplicationHandler : public IApplicationHandler
+{
+public:
+  explicit ApplicationHandler(std::string path = "./bin/applications/");
```
Comment (rBekesh) [inline applications/ExecutionManager/include/application_handler.hpp]: define "./bin/applications/" as a constant in constants.hpp

File: applications/ExecutionManager/include/application_handler.hpp [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+  pid_t startProcess(const ProcessInfo &process) override;
```
Comment (rBekesh) [inline applications/ExecutionManager/include/application_handler.hpp]: (const ProcessInfo& process)

File: applications/ExecutionManager/include/application_handler.hpp [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+   * @brief Method that converts input std::vector of std::string to
+   *        std::vector<char*> to pass as argv in application.
+   * @param vectorToConvert: vector that will be converted.
+   * @return Vector of char* including `nullptr` that be passed to application.
+   */
+    std::vector<char*>
+    convertToNullTerminatingArgv(
+            std::vector<std::string> &vectorToConvert) const;
```
Comment (rBekesh) [inline applications/ExecutionManager/include/application_handler.hpp]: std::vector<std::string>& vectorToConvert) const;

File: applications/ExecutionManager/include/execution_manager.hpp [hunk 16]
```diff
@@ -29,7 +30,8 @@ class ExecutionManager
 {
 public:
 
-  explicit ExecutionManager(std::unique_ptr<IManifestReader> reader);
+  explicit ExecutionManager(std::unique_ptr<IManifestReader> reader,
```
Comment (ulianaStashkiv) [inline applications/ExecutionManager/include/execution_manager.hpp]: explicit not needed here

File: applications/ExecutionManager/src/application_handler.cpp [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+pid_t ApplicationHandler::startProcess(const ProcessInfo &process)
+{
+  pid_t processId = fork();
+
+  if (!processId)
+  {
+    // child process
+    const auto processPath = corePath
+                             + process.createRelativePath();
```
Comment (ulianaStashkiv) [inline applications/ExecutionManager/src/application_handler.cpp]: move + under =

File: applications/ExecutionManager/src/em_main.cpp [hunk 18]
```diff
@@ -8,7 +9,8 @@ int main(int argc, char **argv)
 {
   const char* socketName = "/tmp/execution_management";
```
Comment (ulianaStashkiv) [inline applications/ExecutionManager/src/em_main.cpp]: move to constants.hpp & don't use const char*  use std::string

File: applications/ExecutionManager/src/execution_manager.cpp [hunk 19]
```diff
@@ -24,8 +24,10 @@ const string ExecutionManager::corePath =
 
 const MachineState ExecutionManager::defaultState {"Startup"};
 
-ExecutionManager::ExecutionManager(std::unique_ptr<IManifestReader> reader)
-  : m_activeApplications{}
+ExecutionManager::ExecutionManager(std::unique_ptr<IManifestReader> reader,
+                                   std::unique_ptr<IApplicationHandler> applicationHandler)
+  : appHandler{std::move(applicationHandler)}
+  , m_activeApplications{}
```
Comment (ulianaStashkiv) [inline applications/ExecutionManager/src/execution_manager.cpp]: no leading ,

**Pull request merged**